From a4d9c1705dd657bc3209a7ec1cc92cabeaddd849 Mon Sep 17 00:00:00 2001
From: Ievgen Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Sun, 26 Apr 2020 12:01:14 -0400
Subject: [PATCH 80/93] [MSE] Make SourceBuffer evictCodedFrames less
 aggressive
Source: COMCAST
License: GPLV2
Upstream-Status: None
Signed-off-by: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>

On some streams (with all sync samples and PTS re-odering) evicting
close to current plaback position removes currently playing range
triggering re-buffering.
---
 Source/WebCore/Modules/mediasource/SourceBuffer.cpp | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

Index: git/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
===================================================================
--- git.orig/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
+++ git/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
@@ -988,13 +988,17 @@ void SourceBuffer::evictCodedFrames(size
     MediaTime maximumRangeEnd = currentTime - thirtySeconds;
 
 #if defined(METROLOGICAL)
-    MediaTime microSecond = MediaTime(1, 1000000);
     for (auto& trackBuffer : m_trackBufferMap.values()) {
+        auto t = std::max(MediaTime::zeroTime(), currentTime - MediaTime(2, 1));
         auto prevSync =
-            trackBuffer.samples.decodeOrder().findSyncSamplePriorToPresentationTime(currentTime);
+            trackBuffer.samples.decodeOrder().findSyncSamplePriorToPresentationTime(t);
         if (prevSync != trackBuffer.samples.decodeOrder().rend()) {
             // Don't include the sync frame in the range, just finish right before it.
-            maximumRangeEnd = prevSync->second->presentationTime() - microSecond;
+            prevSync++;
+        }
+        if (prevSync != trackBuffer.samples.decodeOrder().rend()) {
+            maximumRangeEnd = prevSync->second->presentationTime();
+            break;
         }
     }
 #endif
